### ISSUE TITLE
Bug - SLA reporting should honour allow/block list and not monitored entities

### ISSUE THREAD

**Issue by tkoeberl** [guilhemmarchand/trackme#261]
In dashboard: trackMe_qos is not taken care of the block/allow lists OR monitored state.
Does it have a specific reason? 
Why should an SLA be reported on a unmonitored object?

**Comment by guilhemmarchand** [comment]
@tkoeberl 

Thanks a thousands, you are completely right and it should consider only monitored entities.

The new release is upcoming, version 1.2.35, and currently in testing branch:

https://trackme.readthedocs.io/en/testing/releasenotes.html

https://github.com/guilhemmarchand/trackme/tree/testing

I will check this out as I believe the fix will be easy.

**Comment by guilhemmarchand** [comment]
Version 1.2.35 is now live.

**Comment by tkoeberl** [comment]
@guilhemmarchand :
 sure that it is fixed?
![image](https://user-images.githubusercontent.com/16901385/109181460-3861ea80-778c-11eb-9fed-7468ba6e7767.png)

**Comment by guilhemmarchand** [comment]
That should have been @tkoeberl  ;-)

Ok so, I realise there is something missing, technically the allowlist and blocklists are now respected (note: this is refreshed every 4 hours)

There is a new job:

```
TrackMe - gen object summary per object_category
```

This job does the following puts in a KVstore cache a summary of all objects including an isAllowed status that is respecting the allow / blocking lists:

```
| inputlookup trackme_objects_summary
```

Then, the macros rendering the SLA does:

```
| lookup trackme_objects_summary object, object_category OUTPUT isAllowed 
| where (isAllowed == "true")
```

However, what I forgot is the the monitored state should have been excluded too, it's not properly something linked to the allow and block lists but I agree no monitored entities due to the monitored state should not be part of the SLA, I should have added in the SLA macro:

```
[trackme_get_sla(4)]
```

An additional filter on the monitored_state, the definition should be:

patch:

```
| lookup trackme_objects_summary object, object_category OUTPUT isAllowed\
| where isAllowed="true" AND monitored_state="enabled"\
```

Full def:

```
[trackme_get_sla(4)]
definition = `trackme_idx` source="current_state_tracking:$object_category$" object_category="$object_category$" object="$object_freetext$" object="$object$"\
| fields _time object_category, object, current_state\
| stats first(current_state) as current_state by _time, object_category, object\
`comment("#### Use a lookup call to pre-filter the amount of events to streamstats, another lookup call to the same lookup is done at the final which is the most performing scenario #####")`\
| lookup trackme_objects_summary object, object_category OUTPUT isAllowed\
| where isAllowed="true" AND monitored_state="enabled"\
| streamstats last(_time) as "prev_time", last(current_state) as prev_state current=f by object_category, object\
| eval range_duration=_time-prev_time, green_time=case(current_state="green" OR current_state="blue", range_duration), not_green_time=case(current_state!="green" AND current_state!="blue", range_duration)\
| stats sum(range_duration) as range_duration, sum(green_time) as green_time, sum(not_green_time) as not_green_time by object_category, object\
| eval green_time=if(isnum(green_time), green_time, 0), not_green_time=if(isnum(not_green_time), not_green_time, 0), percent_sla=round(green_time/range_duration*100, 2)\
| foreach range_duration green_time not_green_time [ eval <<FIELD>> = tostring('<<FIELD>>', "duration") ]\
| lookup trackme_objects_summary object, object_category OUTPUT priority, monitored_state\
| search priority="$priority$"\
| fields object, object_category, priority, monitored_state, range_duration, green_time, not_green_time, percent_sla\
| rename range_duration as "total duration ([D+]HH:MM:SS)", green_time as "duration spent in green state ([D+]HH:MM:SS)", not_green_time as "duration spent in non green state ([D+]HH:MM:SS)"
args = object_category,object,object_freetext,priority
iseval = 0
```

So my bad, you are right, I will fix it in the next release.

In the mean time you could modify the macro, via Splunk Web the code is:

trackme_get_sla(4)

```
`trackme_idx` source="current_state_tracking:$object_category$" object_category="$object_category$" object="$object_freetext$" object="$object$"
| fields _time object_category, object, current_state
| stats first(current_state) as current_state by _time, object_category, object
`comment("#### Use a lookup call to pre-filter the amount of events to streamstats, another lookup call to the same lookup is done at the final which is the most performing scenario #####")`
| lookup trackme_objects_summary object, object_category OUTPUT isAllowed
| where isAllowed="true"
| streamstats last(_time) as "prev_time", last(current_state) as prev_state current=f by object_category, object
| eval range_duration=_time-prev_time, green_time=case(current_state="green" OR current_state="blue", range_duration), not_green_time=case(current_state!="green" AND current_state!="blue", range_duration)
| stats sum(range_duration) as range_duration, sum(green_time) as green_time, sum(not_green_time) as not_green_time by object_category, object
| eval green_time=if(isnum(green_time), green_time, 0), not_green_time=if(isnum(not_green_time), not_green_time, 0), percent_sla=round(green_time/range_duration*100, 2)
| foreach range_duration green_time not_green_time [ eval <<FIELD>> = tostring('<<FIELD>>', "duration") ]
| lookup trackme_objects_summary object, object_category OUTPUT priority, monitored_state
| search priority="$priority$"
| fields object, object_category, priority, monitored_state, range_duration, green_time, not_green_time, percent_sla
| rename range_duration as "total duration ([D+]HH:MM:SS)", green_time as "duration spent in green state ([D+]HH:MM:SS)", not_green_time as "duration spent in non green state ([D+]HH:MM:SS)"
```

It would become:

```
`trackme_idx` source="current_state_tracking:$object_category$" object_category="$object_category$" object="$object_freetext$" object="$object$"
| fields _time object_category, object, current_state
| stats first(current_state) as current_state by _time, object_category, object
`comment("#### Use a lookup call to pre-filter the amount of events to streamstats, another lookup call to the same lookup is done at the final which is the most performing scenario #####")`
| lookup trackme_objects_summary object, object_category OUTPUT isAllowed
| where isAllowed="true" AND monitored_state="enabled"
| streamstats last(_time) as "prev_time", last(current_state) as prev_state current=f by object_category, object
| eval range_duration=_time-prev_time, green_time=case(current_state="green" OR current_state="blue", range_duration), not_green_time=case(current_state!="green" AND current_state!="blue", range_duration)
| stats sum(range_duration) as range_duration, sum(green_time) as green_time, sum(not_green_time) as not_green_time by object_category, object
| eval green_time=if(isnum(green_time), green_time, 0), not_green_time=if(isnum(not_green_time), not_green_time, 0), percent_sla=round(green_time/range_duration*100, 2)
| foreach range_duration green_time not_green_time [ eval <<FIELD>> = tostring('<<FIELD>>', "duration") ]
| lookup trackme_objects_summary object, object_category OUTPUT priority, monitored_state
| search priority="$priority$"
| fields object, object_category, priority, monitored_state, range_duration, green_time, not_green_time, percent_sla
| rename range_duration as "total duration ([D+]HH:MM:SS)", green_time as "duration spent in green state ([D+]HH:MM:SS)", not_green_time as "duration spent in non green state ([D+]HH:MM:SS)"
```

Thanks for finding my mistakes!

**Comment by tkoeberl** [comment]
@guilhemmarchand : 
everyone has his superpower.. I break things, you fix them.
Thanks a lot for this great app!

**Comment by tkoeberl** [comment]
for me it looks like the filter must be after the 
| lookup trackme_objects_summary object, object_category OUTPUT priority, monitored_state
monitored_state is not there before ;) Means in line:
| search priority="$priority$" monitored_state="enabled"

**Comment by guilhemmarchand** [comment]
@tkoeberl 

I think both would be equal, in the latest version the monitored_state is an output from the lookup call, the reseaon why there is a | search after the where is that the dashboard provides some filtering options, but monitored_state isn't one of the filters, so I am keen to think it would make more sense in the where, what do you think?

You may have noted that there are 2 lookup calls for this search to the same lookup, the reason is performances, after benchmarking it quite a lot in a Prod env, I came to this as the most performing search, the more streamstats is restricted the better ;-)

Thanks a thousands for your help

**Comment by guilhemmarchand** [comment]
@tkoeberl 

Fixed in 1.2.36 published in Splunk Base.
